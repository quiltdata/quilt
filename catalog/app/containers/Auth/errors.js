export class AuthError extends Error {
  static displayName = 'AuthError';
}

export class NotAuthenticated extends AuthError {
  static displayName = 'NotAuthenticated';
}

export class UserAlreadyExists extends AuthError {
  static displayName = 'UserAlreadyExists';

  constructor({ message, ...props }) {
    super(message, props);
  }
}
