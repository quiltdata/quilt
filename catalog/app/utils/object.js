/* object utils */
isEmpty(obj) {
  return Object.keys(obj).length === 0
    && obj.constructor === Object;
    // ^ avoid Object.keys(new Date()).length === 0
}