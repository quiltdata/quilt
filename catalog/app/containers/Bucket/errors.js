import * as R from 'ramda'
import * as React from 'react'
import { connect } from 'react-redux'
import { Route, Link } from 'react-router-dom'
import { createStructuredSelector } from 'reselect'
import Button from '@material-ui/core/Button'

import * as Auth from 'containers/Auth'
import { docs } from 'constants/urls'
import * as NamedRoutes from 'utils/NamedRoutes'
import StyledLink from 'utils/StyledLink'
import { BaseError } from 'utils/error'
import * as RT from 'utils/reactTools'

import Message from './Message'

export class BucketError extends BaseError {}

export class AccessDenied extends BucketError {}

export class CORSError extends BucketError {}

export class NoSuchBucket extends BucketError {}

const WhenAuth = connect(
  createStructuredSelector({
    authenticated: Auth.selectors.authenticated,
  }),
)(({ authenticated, cases, args }) => cases[authenticated](...args))

const whenAuth = (cases) => (...args) => <WhenAuth {...{ cases, args }} />

const SignIn = RT.composeComponent('Bucket.errors.SignIn', () => (
  <NamedRoutes.Inject>
    {({ urls }) => (
      <Route>
        {({ location: l }) => (
          <Button
            component={Link}
            to={urls.signIn(l.pathname + l.search + l.hash)}
            variant="contained"
            color="primary"
          >
            Sign In
          </Button>
        )}
      </Route>
    )}
  </NamedRoutes.Inject>
))

const defaultHandlers = [
  [
    R.is(CORSError),
    () => (
      <Message headline="Error">
        Seems like this bucket is not configured for Quilt.
        <br />
        <StyledLink
          href={`${docs}/references/technical-reference#deploy-a-private-quilt-instance-on-aws`}
        >
          Learn how to configure the bucket for Quilt
        </StyledLink>
        .
      </Message>
    ),
  ],
  [
    R.is(NoSuchBucket),
    () => (
      <Message headline="No Such Bucket">The specified bucket does not exist.</Message>
    ),
  ],
  [
    R.is(AccessDenied),
    whenAuth({
      true: () => (
        <Message headline="Access Denied">
          Seems like you don&apos;t have access to this bucket.
          <br />
          <StyledLink
            href={`${docs}/walkthrough/working-with-the-catalog#access-control`}
          >
            Learn about access control in Quilt
          </StyledLink>
          .
        </Message>
      ),
      false: () => (
        <Message headline="Access Denied">
          Anonymous access not allowed. Please sign in.
          <br />
          <br />
          <SignIn />
        </Message>
      ),
    }),
  ],
]

export const displayError = (pairs = []) =>
  R.cond([
    ...defaultHandlers,
    ...pairs,
    [
      R.T,
      (e) => {
        throw e
      },
    ],
  ])
