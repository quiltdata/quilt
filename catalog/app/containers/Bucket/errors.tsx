import * as R from 'ramda'
import * as React from 'react'
import * as redux from 'react-redux'
import { Route, Link } from 'react-router-dom'
import * as M from '@material-ui/core'

import Message from 'components/Message'
import * as Auth from 'containers/Auth'
import { docs } from 'constants/urls'
import * as NamedRoutes from 'utils/NamedRoutes'
import StyledLink from 'utils/StyledLink'
import { BaseError } from 'utils/error'

export class BucketError extends BaseError {}

export class AccessDenied extends BucketError {}

export class CORSError extends BucketError {}

export class NoSuchBucket extends BucketError {}

interface NoSuchPackageProps {
  bucket: string
  handle: string
}

export class NoSuchPackage extends BucketError {
  static displayName = 'NoSuchPackage'

  constructor(props: NoSuchPackageProps) {
    super(`no package named '${props.handle}' in bucket '${props.bucket}'`, props)
  }
}

export class ESNoIndex extends BucketError {}

export class ESTimeout extends BucketError {}

export class ESThrottle extends BucketError {}

export class FileNotFound extends BucketError {}

export class VersionNotFound extends BucketError {}

export interface BucketPreferencesInvalidProps {
  errors: { instancePath?: string; message?: string }[]
}

export class BucketPreferencesInvalid extends BucketError {
  static displayName = 'BucketPreferencesInvalid'

  constructor(props: BucketPreferencesInvalidProps) {
    super(
      props.errors
        .map(({ instancePath, message }) => `${instancePath} ${message}`)
        .join(', '),
      props,
    )
  }
}

export interface WorkflowsConfigInvalidProps {
  errors: { instancePath?: string; message?: string }[]
}

export class WorkflowsConfigInvalid extends BucketError {
  static displayName = 'WorkflowsConfigInvalid'

  constructor(props: WorkflowsConfigInvalidProps) {
    super(
      props.errors
        .map(({ instancePath, message }) =>
          instancePath ? `${instancePath} ${message}` : message,
        )
        .join(', '),
      props,
    )
  }
}

export interface ManifestTooLargeProps {
  bucket: string
  key: string
  actualSize: number
  maxSize: number
}

export class ManifestTooLarge extends BucketError {
  static displayName = 'ManifestTooLarge'

  constructor(props: ManifestTooLargeProps) {
    super(
      `Package manifest at s3://${props.bucket}/${props.key} is too large: ${props.actualSize} (max size: ${props.maxSize})`,
      props,
    )
  }
}

export interface BadRevisionProps {
  bucket: string
  handle: string
  revision: string
}

export class BadRevision extends BucketError {
  static displayName = 'BadRevision'

  constructor(props: BadRevisionProps) {
    super(
      `Could not resolve revision "${props.revision}" for package "${props.handle}" in s3://${props.bucket}`,
      props,
    )
  }
}

interface WhenAuthCases {
  true: () => React.ReactElement
  false: () => React.ReactElement
}

interface WhenAuthProps {
  cases: WhenAuthCases
}

function WhenAuth({ cases }: WhenAuthProps) {
  const authenticated = redux.useSelector(Auth.selectors.authenticated)
  return cases[`${authenticated}` as 'true' | 'false']()
}

const whenAuth = (cases: WhenAuthCases) => () => <WhenAuth cases={cases} />

function SignIn() {
  const { urls } = NamedRoutes.use()
  return (
    <Route>
      {({ location: l }) => (
        <M.Button
          component={Link}
          to={urls.signIn(l.pathname + l.search + l.hash)}
          variant="contained"
          color="primary"
        >
          Sign In
        </M.Button>
      )}
    </Route>
  )
}

const defaultHandlers: ErrorHandler[] = [
  [
    R.is(CORSError),
    () => (
      <Message headline="Error">
        Seems like this bucket is not configured for Quilt.
        <br />
        <StyledLink
          target="_blank"
          href={`${docs}/references/technical-reference#deploy-a-private-quilt-instance-on-aws`}
        >
          Learn how to configure the bucket for Quilt
        </StyledLink>
        .
      </Message>
    ),
  ],
  [
    R.is(NoSuchBucket),
    () => (
      <Message headline="No Such Bucket">The specified bucket does not exist.</Message>
    ),
  ],
  [
    R.is(NoSuchPackage),
    (e: NoSuchPackage) => (
      <Message headline="No Such Package">
        Package named{' '}
        <M.Box component="span" fontWeight="fontWeightMedium">{`"${e.handle}"`}</M.Box>{' '}
        could not be found in this bucket.
      </Message>
    ),
  ],
  [
    R.is(ESNoIndex),
    () => (
      <Message headline="Oops, no search cluster">
        The bucket owner needs to{' '}
        <StyledLink target="_blank" href={docs}>
          tie this bucket to Quilt
        </StyledLink>{' '}
        to enable Packages, Search, and detailed Overviews.
      </Message>
    ),
  ],
  [
    R.either(R.is(ESTimeout), R.is(ESThrottle)),
    () => (
      <Message headline="Whoa, search cluster is stressed out">
        ElasticSearch is stressed out. Please try again in a few minutes.
      </Message>
    ),
  ],
  [
    R.is(AccessDenied),
    whenAuth({
      true: () => (
        <Message headline="Access Denied">
          Seems like you don&apos;t have access to this bucket.
          <br />
          <StyledLink
            target="_blank"
            href={`${docs}/walkthrough/working-with-the-catalog#access-control`}
          >
            Learn about access control in Quilt
          </StyledLink>
          .
        </Message>
      ),
      false: () => (
        <Message headline="Access Denied">
          Anonymous access not allowed. Please sign in.
          <br />
          <br />
          <SignIn />
        </Message>
      ),
    }),
  ],
]

type ErrorHandler = [(error: unknown) => boolean, (error: any) => React.ReactElement]

export const displayError = (pairs: ErrorHandler[] = []) =>
  R.cond([
    ...defaultHandlers,
    ...pairs,
    [
      R.T,
      (e) => {
        throw e
      },
    ],
  ])

const startsWithSafe = (prefix: string) => (str: unknown) =>
  typeof str === 'string' ? str.startsWith(prefix) : false

type ErrorCatcher = [(error: unknown) => boolean, (error: unknown) => never]

export const catchErrors = (pairs: ErrorCatcher[] = []) =>
  R.cond([
    [
      R.propEq('message', 'Network Failure'),
      () => {
        throw new CORSError()
      },
    ],
    [
      R.propEq('message', 'Access Denied'),
      () => {
        throw new AccessDenied()
      },
    ],
    [
      R.propEq('code', 'Forbidden'),
      () => {
        throw new AccessDenied()
      },
    ],
    [
      R.propEq('code', 'NoSuchBucket'),
      () => {
        throw new NoSuchBucket()
      },
    ],
    [
      R.propEq(
        'message',
        "API Gateway error (500): NotFoundError(404, 'index_not_found_exception', 'no such index')",
      ),
      () => {
        throw new ESNoIndex()
      },
    ],
    [
      R.propSatisfies(
        startsWithSafe(
          'API Gateway error (500): ConnectionTimeout caused by - ReadTimeout',
        ),
        'message',
      ),
      () => {
        throw new ESTimeout()
      },
    ],
    [
      R.propSatisfies(
        startsWithSafe(
          "API Gateway error (500): AuthorizationException(403, '403 Request throttled due to too many requests",
        ),
        'message',
      ),
      () => {
        throw new ESThrottle()
      },
    ],
    ...pairs,
    [
      R.T,
      (e) => {
        throw e
      },
    ],
  ]) as (reason: unknown) => never
