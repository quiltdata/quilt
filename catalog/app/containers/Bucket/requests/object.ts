import type { S3 } from 'aws-sdk'

import * as quiltConfigs from 'constants/quiltConfigs'
import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'
import type { JsonRecord } from 'utils/types'
import * as workflows from 'utils/workflows'

import { FileNotFound, VersionNotFound } from '../errors'

import { decodeS3Key } from './utils'
import { ensureObjectIsPresent } from './requestsUntyped'

interface ObjectTagsArgs {
  s3: S3
  handle: Model.S3.S3ObjectLocation
}

export const objectTags = ({
  s3,
  handle: { bucket, key, version },
}: ObjectTagsArgs): Promise<Record<string, string>> =>
  s3
    .getObjectTagging({
      Bucket: bucket,
      Key: key,
      VersionId: version,
    })
    .promise()
    .then(({ TagSet }) =>
      TagSet.reduce((memo, { Key, Value }) => ({ ...memo, [Key]: Value }), {}),
    )

interface ObjectMetaArgs {
  s3: S3
  handle: Model.S3.S3ObjectLocation
}

export const objectMeta = async ({
  s3,
  handle: { bucket, key, version },
}: ObjectMetaArgs): Promise<JsonRecord | undefined> => {
  const r = await s3
    .headObject({
      Bucket: bucket,
      Key: key,
      VersionId: version,
    })
    .promise()
  if (r.Metadata?.helium) {
    return JSON.parse(r.Metadata?.helium)
  }
}

type ExistingObject = Model.S3.S3ObjectLocation & { size?: number; lastModified?: Date }

interface EnsureObjectIsPresentInCollectionArgs {
  s3: S3
  handles: Model.S3.S3ObjectLocation[]
}

const ensureObjectIsPresentInCollection = async ({
  s3,
  handles,
}: EnsureObjectIsPresentInCollectionArgs): Promise<ExistingObject | null> => {
  if (!handles.length) return null

  const [handle, ...handlesTail] = handles
  const existingObject = await ensureObjectIsPresent({
    s3,
    ...handle,
  })

  return (
    existingObject ||
    (await ensureObjectIsPresentInCollection({ s3, handles: handlesTail }))
  )
}

interface GetObjectArgs {
  s3: S3
  handle: Model.S3.S3ObjectLocation
}

const getObject = ({ s3, handle }: GetObjectArgs) =>
  s3
    .getObject({
      Bucket: handle.bucket,
      Key: handle.key,
      VersionId: handle.version,
    })
    .promise()
    .then(({ Body }) => ({
      handle,
      body: Body,
    }))

interface ObjectVersionsArgs {
  s3: S3
  bucket: string
  path: string
}

type ListItem = S3.ObjectVersion | S3.DeleteMarkerEntry

const isDeleteMarker = (v: ListItem): v is S3.DeleteMarkerEntry =>
  (v as S3.ObjectVersion).Size == null

const isObjectVersion = (v: ListItem): v is S3.ObjectVersion =>
  (v as S3.ObjectVersion).Size != null

export const objectVersions = async ({ s3, bucket, path }: ObjectVersionsArgs) => {
  const { Versions, DeleteMarkers } = await s3
    .listObjectVersions({ Bucket: bucket, Prefix: path, EncodingType: 'url' })
    .promise()
  return ([...(Versions || []), ...(DeleteMarkers || [])] as ListItem[])
    .filter(({ Key }) => decodeS3Key(Key || '') === path)
    .map((v) => ({
      isLatest: v.IsLatest || false,
      lastModified: v.LastModified,
      // TODO: make two separate maps, object version can be without `Size`
      size: isObjectVersion(v) ? v.Size : undefined,
      id: v.VersionId,
      deleteMarker: isDeleteMarker(v),
      archived: isObjectVersion(v)
        ? v.StorageClass === 'GLACIER' || v.StorageClass === 'DEEP_ARCHIVE'
        : false,
    }))
    .toSorted(({ lastModified: left }, { lastModified: right }) => {
      if (left && right) return right.getTime() - left.getTime()
      if (left) return -1
      if (right) return 1
      return 0
    })
}

interface FetchFileInCollectionArgs {
  s3: S3
  handles: Model.S3.S3ObjectLocation[]
}

export async function fetchFileInCollection({ s3, handles }: FetchFileInCollectionArgs) {
  const existingObject = await ensureObjectIsPresentInCollection({ s3, handles })
  if (!existingObject) {
    throw new FileNotFound(`No object in ${JSON.stringify(handles)} exist`)
  }
  return getObject({ s3, handle: existingObject })
}

interface FetchFile {
  s3: S3
  handle: Model.S3.S3ObjectLocation
}

export async function fetchFile({ s3, handle }: FetchFile) {
  const existingObject: ExistingObject | null = await ensureObjectIsPresent({
    s3,
    ...handle,
  })
  if (!existingObject) {
    throw new FileNotFound(`Object ${JSON.stringify(handle)} doesn't exist`)
  }
  return getObject({ s3, handle: existingObject })
}

interface MetadataSchemaArgs {
  s3: S3
  schemaUrl?: string
}

export const metadataSchema = async ({ s3, schemaUrl }: MetadataSchemaArgs) => {
  if (!schemaUrl) return null

  const handle = s3paths.parseS3Url(schemaUrl)

  const response = await fetchFile({ s3, handle })
  return JSON.parse(response.body?.toString('utf-8') || '{}')
}

export const WORKFLOWS_CONFIG_PATH = quiltConfigs.workflows
// TODO: enable this when backend is ready
// const WORKFLOWS_CONFIG_PATH = [
//   '.quilt/workflows/config.yaml',
//   '.quilt/workflows/config.yml',
// ]

interface WorkflowsConfigArgs {
  s3: S3
  bucket: string
}

export const workflowsConfig = async ({ s3, bucket }: WorkflowsConfigArgs) => {
  try {
    const response = await fetchFile({
      s3,
      handle: { bucket, key: WORKFLOWS_CONFIG_PATH },
    })
    return workflows.parse(response.body?.toString('utf-8') || '')
  } catch (e) {
    if (e instanceof FileNotFound || e instanceof VersionNotFound)
      return workflows.emptyConfig

    // eslint-disable-next-line no-console
    console.log('Unable to fetch')
    // eslint-disable-next-line no-console
    console.error(e)
    throw e
  }
}
