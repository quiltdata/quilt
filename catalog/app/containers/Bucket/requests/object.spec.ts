import type { S3 } from 'aws-sdk'

import { FileNotFound } from '../errors'

import { fetchFile, objectVersions } from './object'

class AWSError extends Error {
  code: string

  constructor(code: string, message?: string) {
    super(message)
    this.code = code
  }
}

jest.mock(
  'constants/config',
  jest.fn(() => ({})),
)

describe('app/containers/Bucket/requests/object', () => {
  describe('objectVersions', () => {
    const s3 = {
      listObjectVersions: () => ({
        promise: () =>
          Promise.resolve({
            Versions: [
              {
                Key: 'foo',
                LastModified: new Date(1732811111111),
                Size: 1,
                VersionId: 'earlier date later',
              },
              { Key: 'foo', Size: 1, VersionId: 'essential' },
              { Key: 'foo', Size: 1 }, // without version
              { Key: 'foo', IsLatest: true, VersionId: 'latest', Size: 1 },
              {
                Key: 'foo',
                Size: 1,
                StorageClass: 'GLACIER',
                VersionId: 'archived glacier',
              },
              {
                Key: 'foo',
                Size: 1,
                StorageClass: 'DEEP_ARCHIVE',
                VersionId: 'deep archived',
              },
              {
                Key: 'foo',
                LastModified: new Date(1732855555555),
                Size: 1,
                VersionId: 'later date first',
              },
              { Key: 'foo', VersionId: 'marked as deleted because no `Size`' },
              { Key: 'drop' },
            ],
            DeleteMarkers: [
              { Key: 'foo', VersionId: 'deleted' },
              {
                Key: 'foo',
                LastModified: new Date(1732899999999),
                VersionId: 'deleted, but the most recent',
              },
            ],
          } as S3.Types.ListObjectVersionsOutput),
      }),
    }
    it('return object versions', () =>
      expect(
        objectVersions({ s3: s3 as S3, bucket: 'any', path: 'foo' }),
      ).resolves.toMatchSnapshot())
  })

  describe('fetchFile', () => {
    const s3 = {
      getObject: () => ({
        promise: () =>
          Promise.resolve({
            Body: Buffer.from('{"foo": "bar"}'),
          } as S3.Types.GetObjectOutput),
      }),
      headObject: ({ Key }: S3.Types.HeadObjectRequest) => ({
        // AWS Request has no `.response` field type
        // but it has this in practice, and we use it
        response: { httpResponse: { headers: {} } },
        promise: () => {
          switch (Key) {
            case 'does-not-exist':
              return Promise.reject(new AWSError('NotFound'))
            case 'exist':
              return Promise.resolve({
                VersionId: 'resolved',
              } as S3.Types.HeadObjectOutput)
            default:
              return Promise.reject(new Error())
          }
        },
      }),
    }

    it('fetches existing file', async () => {
      const result = await fetchFile({
        // `.response` is absent in type
        // @ts-expect-error
        s3: s3 as S3,
        handle: { bucket: 'b', key: 'exist' },
      })
      expect(result.handle).toMatchObject({
        bucket: 'b',
        key: 'exist',
        version: 'resolved',
      })
      expect(result.body?.toString()).toBe('{"foo": "bar"}')
    })

    it('throws when file not found', async () => {
      const result = fetchFile({
        // `.response` is absent in type
        // @ts-expect-error
        s3: s3 as S3,
        handle: { bucket: 'b', key: 'does-not-exist' },
      })
      return expect(result).rejects.toThrow(FileNotFound)
    })

    it('re-throws on error', async () => {
      const result = fetchFile({
        // `.response` is absent in type
        // @ts-expect-error
        s3: s3 as S3,
        handle: { bucket: 'b', key: 'error' },
      })
      return expect(result).rejects.toThrow(Error)
    })
  })
})
