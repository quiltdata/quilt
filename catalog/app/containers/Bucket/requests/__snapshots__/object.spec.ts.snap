// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/containers/Bucket/requests/object objectVersions return object versions 1`] = `
[
  {
    "archived": false,
    "deleteMarker": true,
    "id": "deleted, but the most recent",
    "isLatest": false,
    "lastModified": 2024-11-29T17:06:39.999Z,
    "size": undefined,
  },
  {
    "archived": false,
    "deleteMarker": false,
    "id": "later date first",
    "isLatest": false,
    "lastModified": 2024-11-29T04:45:55.555Z,
    "size": 1,
  },
  {
    "archived": false,
    "deleteMarker": false,
    "id": "earlier date later",
    "isLatest": false,
    "lastModified": 2024-11-28T16:25:11.111Z,
    "size": 1,
  },
  {
    "archived": false,
    "deleteMarker": false,
    "id": "essential",
    "isLatest": false,
    "lastModified": undefined,
    "size": 1,
  },
  {
    "archived": false,
    "deleteMarker": false,
    "id": undefined,
    "isLatest": false,
    "lastModified": undefined,
    "size": 1,
  },
  {
    "archived": false,
    "deleteMarker": false,
    "id": "latest",
    "isLatest": true,
    "lastModified": undefined,
    "size": 1,
  },
  {
    "archived": true,
    "deleteMarker": false,
    "id": "archived glacier",
    "isLatest": false,
    "lastModified": undefined,
    "size": 1,
  },
  {
    "archived": true,
    "deleteMarker": false,
    "id": "deep archived",
    "isLatest": false,
    "lastModified": undefined,
    "size": 1,
  },
  {
    "archived": false,
    "deleteMarker": true,
    "id": "marked as deleted because no \`Size\`",
    "isLatest": false,
    "lastModified": undefined,
    "size": undefined,
  },
  {
    "archived": false,
    "deleteMarker": true,
    "id": "deleted",
    "isLatest": false,
    "lastModified": undefined,
    "size": undefined,
  },
]
`;
